File: problems-extend/0901-1000/0948_bag-of-tokens/solution.java
import java.util.Arrays;

class Solution {

    public int bagOfTokensScore(int[] tokens, int power) {
        // An optimal plan buys points with the cheapest tokens and sells the
        // dearest ones for power, so sort and walk two pointers inward.
        Arrays.sort(tokens);
        int left = 0, right = tokens.length - 1;
        int score = 0, best = 0;
        while (left <= right) {
            if (power >= tokens[left]) {
                // Affordable: buy a point with the cheapest remaining token.
                power -= tokens[left];
                score++;
                left++;
                best = Math.max(best, score);
            } else if (score >= 1 && left < right) {
                // Broke: sell a point for the power of the dearest token,
                // keeping one token in play to spend it on.
                power += tokens[right];
                score--;
                right--;
            } else {
                break;
            }
        }
        return best;
    }
}

File: problems-extend/0901-1000/0948_bag-of-tokens/solution.rs
impl Solution {
    pub fn bag_of_tokens_score(tokens: Vec<i32>, power: i32) -> i32 {
        // An optimal plan buys points with the cheapest tokens and sells the
        // dearest ones for power, so sort and walk two pointers inward.
        let mut tokens = tokens;
        let mut power = power;
        tokens.sort();
        let mut left: isize = 0;
        let mut right: isize = tokens.len() as isize - 1;
        let mut score = 0;
        let mut best = 0;
        while left <= right {
            if power >= tokens[left as usize] {
                // Affordable: buy a point with the cheapest remaining token.
                power -= tokens[left as usize];
                score += 1;
                left += 1;
                best = best.max(score);
            } else if score >= 1 && left < right {
                // Broke: sell a point for the power of the dearest token,
                // keeping one token in play to spend it on.
                power += tokens[right as usize];
                score -= 1;
                right -= 1;
            } else {
                break;
            }
        }
        best
    }
}
